[Ryan Tan] iP

ryan-txn opened the pull request.

ryan-txn committed:
replace Duke file with Tommi file

ryan-txn committed:
Revert "replace Duke file with Tommi file"

This reverts commit 812d6121ab6ced64ef898ea73826b493f640947a.

ryan-txn committed:
Reapply "replace Duke file with Tommi file"

This reverts commit 5f645dcbe6f9038037d8c0bc08bc78371f91e52a.

ryan-txn committed:
Add repeating strings and detecting "bye" string

ryan-txn committed:
Add list of tasks

Add functions to print the intro and exit messages, add elements to the list and print the list.

ryan-txn committed:
Add checklist

ryan-txn committed:
Fix coding standards

LTK-1606 commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+            } else if (input.equals("list")) {
+                listTasks();
+            } else if (input.startsWith("mark ")) {
+                int taskIndex = Integer.parseInt(input.substring(5)) - 1;
+                markTask(taskIndex);
+            } else if (input.startsWith("unmark ")) {
+                int taskIndex = Integer.parseInt(input.substring(7)) - 1;
+                unmarkTask(taskIndex);
```
Consider not having multiple levels of abstraction in a single code fragment

LTK-1606 commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+        Scanner scanner = new Scanner(System.in);
+        String input;
+
+        while (true) {
```
Maybe consider using a variable instead for readability

LTK-1606 commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+
+public class Tommi {
+    private static final int MAX_TASKS = 100;
+    private static final String[] tasks = new String[MAX_TASKS];  // Array to store tasks
+    private static final boolean[] isCompleted = new boolean[MAX_TASKS];  // Array to store task completion status
+    private static int taskCount = 0;  // Counter to keep track of the number of tasks
```
Why not use a Class to represent all instances of tasks?

3CCLY replied:
I concur

LTK-1606 commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private static void markTask(int index) {
+        if (index >= 0 && index < taskCount) {
+            isCompleted[index] = true;
+            printLine();
+            System.out.println("Awesomesauce! I've marked this task as done:");
+            System.out.println("  [X] " + tasks[index]);
+            printLine();
+        }
```
Consider including error handling

LTK-1606 commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private static void printLine() {
+        System.out.println("____________________________________________________________");
```
Consider using ( "_*30" ) instead

LTK-1606 reviewed (commented):
Overall, looks pretty ok! Incorporating Error Handling and Classes could save you a lot of headache in the future

Zackermax commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private static void printIntroMessage() {
+        String intro =
+                          " ______                   \n"
+                        + "/_  __/__  __ _  __ _  (_)\n"
+                        + " / / / _ \\/  ' \\/  ' \\/ / \n"
+                        + "/_/  \\___/_/_/_/_/_/_/_/  \n"
+                        + "____________________________________________________________\n"
+                        + "Hello! I'm Tommi!\n"
+                        + "What can I do for you?\n"
+                        + "____________________________________________________________\n";
```
Indentation for wrapped lines should be 8 spaces (i.e. twice the normal indentation of 4 spaces) more than the parent line.

3CCLY replied:
You can consider using 8 spaces to split up your line as stated in the coding standard

Zackermax commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+        while (true) {
+            input = scanner.nextLine(); // Read user input
+
+            if (input.equals("bye")) {
+                printExitMessage();
+                break; // Exit the loop
+            } else if (input.equals("list")) {
+                listTasks();
+            } else if (input.startsWith("mark ")) {
+                int taskIndex = Integer.parseInt(input.substring(5)) - 1;
+                markTask(taskIndex);
+            } else if (input.startsWith("unmark ")) {
+                int taskIndex = Integer.parseInt(input.substring(7)) - 1;
+                unmarkTask(taskIndex);
+            } else {
+                addTask(input);
+            }
```
Code would be much neater with a switch-case structure instead of if-else.

3CCLY commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+        if (index >= 0 && index < taskCount) {
+            isCompleted[index] = true;
+            printLine();
+            System.out.println("Awesomesauce! I've marked this task as done:");
```
I like the personality added to the chatbot

Zackermax commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+    private static void listTasks() {
+        printLine();
+        System.out.println("Here are the tasks in your list:");
+        for (int i = 0; i < taskCount; i++) {
+            String status = isCompleted[i] ? "[X]" : "[ ]";
+            System.out.println((i + 1) + "." + status + " " + tasks[i]);
+        }
+        printLine();
+    }
+
+    private static void markTask(int index) {
+        if (index >= 0 && index < taskCount) {
+            isCompleted[index] = true;
+            printLine();
+            System.out.println("Awesomesauce! I've marked this task as done:");
+            System.out.println("  [X] " + tasks[index]);
+            printLine();
+        }
+    }
+
+    private static void unmarkTask(int index) {
+        if (index >= 0 && index < taskCount) {
+            isCompleted[index] = false;
+            printLine();
+            System.out.println("OK, I've marked this task as undone:");
+            System.out.println("  [ ] " + tasks[index]);
+            printLine();
+        }
+    }
+
+    private static void printExitMessage() {
+        printLine();
+        System.out.println("Bye. Hope to see you again soon!");
+        printLine();
+    }
+
+    private static void printLine() {
+        System.out.println("____________________________________________________________");
+    }
+}
```
The code is good overall. It avoids long methods, deep nesting, complicated expressions and magic numbers. The code is well structured and easily understood.

3CCLY commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Scanner;
```
You can consider adding comments to help with readability

3CCLY reviewed (commented):
Looks good to me in general, just some nitpick

ryan-txn committed:
Employ OOP Princples

Use encapsulation to encapsulate input handler. Implement function abstraction in Tommi.java

ryan-txn committed:
Include ToDo, Deadline and Event Task subclasses

Yvorm commented on src/main/java/Tommi.java:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    private static void printIntroMessage() {
+        String intro = """
+                 ______                  \s
+                /_  __/__  __ _  __ _  (_)
+                 / / / _ \\/  ' \\/  ' \\/ /\s
+                /_/  \\___/_/_/_/_/_/_/_/ \s
+                ____________________________________________________________
+                Hello! I'm Tommi!
+                What can I do for you?
+                ____________________________________________________________
+                """;
+        System.out.println(intro);
+    }
```
I appreciate the effort gone to customise your logo, and abstract it to a function for later use!

Yvorm commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,18 @@
+public class Deadline extends Task{
+
+    protected String by;
+
+    public Deadline(boolean isDone, String taskName, String by) {
+        super(isDone, taskName);
+        this.by = by;
+    }
+
+    public Deadline updateIsDone(boolean newIsDone) {
+        return new Deadline(newIsDone, this.taskName, this.by);
+    }
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by + ")";
+    }
+}
```
I like the immutability of the updateIsDone method and the clear toString representation. A small suggestion would be to validate the by field if it's supposed to follow a specific format (like a date or time). You might also want to add a method to update the deadline (by) in the future, just like you did for isDone. Otherwise, everything looks really well-structured!

Yvorm commented on src/main/java/InputHandler.java:
```diff
@@ -0,0 +1,40 @@
+public class InputHandler {
+
+    public static void processInputCases(String input) {
+        if (input.equals("list")) {
+            TaskList.listTasks();
+            return;
+        }
+
+        String[] words = input.split(" ");
+        int taskIndex;
+
+        switch (words[0]) {
+        case "mark":
+            taskIndex = Integer.parseInt(words[1]) - 1;
+            TaskList.markTask(taskIndex);
+            break;
+        case "unmark":
+            taskIndex = Integer.parseInt(input.substring(7)) - 1;
+            TaskList.unmarkTask(taskIndex);
+            break;
+        case "todo":
+            TaskList.addTask(new ToDo(false, input.substring(5)));
+            break;
+        case "deadline":
+            String[] deadlineParts = input.substring(9).split("/by", 2);
+            TaskList.addTask(new Deadline(false, deadlineParts[0], deadlineParts[1]));
+            break;
```
I like the clear logic, use of switch and how compact everything is! One thing to consider is adding more error handling, especially around Integer.parseInt() and array indexing, to prevent crashes on invalid inputs. It might also be worth refactoring the splitting of strings into helper methods for better readability .

Yvorm commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,46 @@
+public class TaskList {
+
+    private static final int MAX_TASKS = 100;
+
+    private static final Task[] tasks = new Task[MAX_TASKS];  // Array to store tasks
+    private static int tasksCount = 0;  // Counter to keep track of the number of tasks
+
+    public static void listTasks() {
+        Tommi.printLine();
+        System.out.println("Here are the tasks in your list:");
+        for (int i = 0; i < tasksCount; i++) {
+            System.out.println((i + 1) + ". "  + tasks[i]);
+        }
+        Tommi.printLine();
+    }
+
+    public static void addTask(Task task) {
+        tasks[tasksCount] = task;
+        tasksCount++;
+        Tommi.printLine();
+        System.out.println("Sure. I've added the task: " + System.lineSeparator()
+                + task + System.lineSeparator()
+                + "There are now " + tasksCount + " tasks in the list.");
+        Tommi.printLine();
```
The class is well structured, and the methods are concise and really easy to understand 😄. Would you want to switch a dynamic data structure later down the track (i.e. ArrayList), to give you some scalability? It might also be good to add boundary checks for adding tasks to ensure it doesn’t exceed MAX_TASKS. Overall, great foundation!

Yvorm reviewed (commented):
Overall, I think this is very concise and readable. The logic used is straightforward and effective too. In some places I have nits regarding edge case inputs which may crash the program, but according to the project specifications everything has been done very well!

ryan-txn committed:
Handle exceptions

Print error message for input with no task type, list or bye, and throw exception when doing so.

ryan-txn committed:
Revert "Handle exceptions"

This reverts commit 85d21ebbb09d8b617168b546f8576840fe43d270.

ryan-txn committed:
Reapply "Handle exceptions"

This reverts commit baaa89231597e861b5a7e4d15e2b2d5c8213fda4.

ryan-txn committed:
Revert "Handle exceptions"

This reverts commit 85d21ebbb09d8b617168b546f8576840fe43d270.

ryan-txn committed:
Merge branch 'master' into branch-Level-5

ryan-txn committed:
Merge branch 'branch-Level-5'

ryan-txn committed:
Redo exception handling, Package code

ryan-txn committed:
add commit for branch

adding a change to level 5 branch so it branches after updating it to main

ryan-txn committed:
Revert "add commit for branch"

This reverts commit 792321a8c226e386b0511368f254b8a9190029b2.

wenxin-c commented on src/main/java/tommi/InputHandler.java:
```diff
@@ -0,0 +1,58 @@
+                TaskList.markTask(taskIndex);
+                break;
+            case "unmark":
+                taskIndex = Integer.parseInt(input.substring(7)) - 1;
```
Try to avoid magic literals(e.g. numbers, strings), you can consider using named constants to give the literals a meaning so that it will be easier to understand. 
E.g.
```
static final double PI = 3.14236;
static final int MAX_SIZE = 10;
...
return PI;
...
return MAX_SIZE - 1;

```

wenxin-c commented on src/main/java/tommi/TaskList.java:
```diff
@@ -0,0 +1,48 @@
+        tasksCount++;
+        Tommi.printLine();
+        System.out.println("Sure. I've added the task: " + System.lineSeparator()
+                + task + System.lineSeparator()
+                + "There are now " + tasksCount + " tasks in the list.");
```
Please be consistent in placing the + sign. In ```InputHandler.java``` file you have been putting the + sign at the back when you wrap lines. You can consider following course website e.g. 
```
setText("Long line split"
        + "into two parts.");
if (isReady) {
    setText("Long line split"
            + "into two parts.");
}

```

wenxin-c commented on src/main/java/tommi/TaskList.java:
```diff
@@ -0,0 +1,48 @@
+        Tommi.printLine();
+    }
+
+    public static void markTask(int index) {
+        if (index >= 0 && index < tasksCount) {
+            Tommi.printLine();
+            System.out.println("Awesomesauce! I've marked this task as done:");
+            tasks[index] = tasks[index].updateIsDone(true);
+            System.out.println(tasks[index]);
+            Tommi.printLine();
+        }
+    }
+
+    public static void unmarkTask(int index) {
+        if (index >= 0 && index < tasksCount) {
+            Tommi.printLine();
+            System.out.println("OK, I've marked this task as undone:");
+            tasks[index] = tasks[index].updateIsDone(false);
+            System.out.println(tasks[index]);
+            Tommi.printLine();
+        }
+    }
```
You can consider extracting these two functions further and pass in variables since they are pretty similar.

wenxin-c reviewed (commented):
Overall good job in coding standard and code quality, just a few points for you to consider.

ryan-txn committed:
Use ArrayList to replace Array and Add delete task function

ryan-txn committed:
Merge branch 'branch-Level-6'

ryan-txn committed:
Add save and load functions

ryan-txn committed:
Merge branch 'branch-Level-7'

ryan-txn committed:
Rename classes and add Task package

ryan-txn committed:
Add Search function

ryan-txn committed:
Improve readability and Shift readInputStrings method from Ui to Parser

ryan-txn committed:
Move all print functions to Ui class

ryan-txn committed:
Improve OOP

ryan-txn committed:
Merge pull request #1 from ryan-txn/branch-Level-9

Branch level 9

ryan-txn committed:
JavaDoc comments

Make Task an abstract class

ryan-txn committed:
Merge pull request #2 from ryan-txn/heads/A-JavaDoc

JavaDoc comments

ryan-txn committed:
Update User Guide

Minor code touch ups

ryan-txn committed:
Merge pull request #3 from ryan-txn/heads/A-UserGuide

Update User Guide

ryan-txn committed:
Save file bug fix

Handle first time opening file error caused by new save file not created
